File: src/main/java/com/goblinskeep/entity/Goblin.java
package com.goblinskeep.entity;

import java.awt.*;
import com.goblinskeep.app.Direction;
import com.goblinskeep.app.GamePanel;
import com.goblinskeep.pathFinder.Circle;

/**
 * Represents a goblin enemy in the game.
 * This is an abstract class that defines common goblin behavior and appearance.
 * Goblins are hostile NPCs that interact with the player and game environment.
 */
public abstract class Goblin extends Entity{
    /** Reference to the player, used for AI movement and interaction. */
    protected Player player;

    /** The direction in which the goblin is currently facing/drawn. */
    protected Direction drawDirection = Direction.UP;

    /** Indicates whether the goblin is currently following a path to the player. */
    public boolean onPath;

    /** Indicates whether the goblin currently has line of sight to the player. */
    public boolean inSight;

    /** Line-of-sight radius in world pixels. The goblin starts chasing when the player's center enters this circle. */
    public int LOSradius = 120;

    /** Manhattan tile distance beyond which a chasing goblin gives up if it has lost line of sight. */
    private static final int DISENGAGE_TILE_DISTANCE = 10;

    /**
     * Constructs a goblin with references to the game panel and player.
     *
     * @param gp     The game panel instance.
     * @param player The player instance.
     * @param WorldX the Goblin X coordinate in the world
     * @param WorldY the Goblin Y coordinate in the world
     */
    public Goblin(GamePanel gp, Player player, int WorldX, int WorldY) {
        super(gp, WorldX, WorldY);  // Pass values up to Entity constructor
        // this.gp = gp;
        this.player = player;
        this.speed = 2;
        this.collisionArea = new Rectangle(11, 17, 23, 23); // Set collision area
        this.hitboxDefaultX = collisionArea.x;
        this.hitboxDefaultY = collisionArea.y;
        direction = Direction.DOWN;
        getGoblinImage();
    }


    /**
     * Defines the goblin's behavior (to be implemented by subclasses).
     * This method should contain movement and AI logic.
     */
    public abstract void getAction();


    /**
     * Loads (using loadImage) and assigns sprite images for goblin animations in different directions.
     * Called when initializing the goblin's appearance.
     */
    public void getGoblinImage(){
        up1    = Entity.loadImage("/goblin/orc_up_1.png");
        up2    = Entity.loadImage("/goblin/orc_up_2.png");
        down1  = Entity.loadImage("/goblin/orc_down_1.png");
        down2  = Entity.loadImage("/goblin/orc_down_2.png");
        right1 = Entity.loadImage("/goblin/orc_right_1.png");
        right2 = Entity.loadImage("/goblin/orc_right_2.png");
        left1  = Entity.loadImage("/goblin/orc_left_1.png");
        left2  = Entity.loadImage("/goblin/orc_left_2.png");
    }
    
    /**
     * Updates the goblin's state. Recomputes line-of-sight, latches the chase state on
     * when the player enters LOS, latches it off only when LOS is lost AND the player is
     * more than {@link #DISENGAGE_TILE_DISTANCE} tiles away (so the goblin keeps pursuing
     * briefly after a corner). Then delegates to the subclass's {@link #getAction()}.
     */
    public void update(){
        int xDistance = Math.abs(WorldX - gp.Player.WorldX);
        int yDistance = Math.abs(WorldY - gp.Player.WorldY);
        int tileDistance = (xDistance + yDistance) / gp.tileSize;

        checkLOS();

        if (!onPath && inSight) {
            onPath = true;
        }
        if (onPath && !inSight && tileDistance > DISENGAGE_TILE_DISTANCE) {
            onPath = false;
        }

        getAction();
    }

    /**
     * Sets {@link #inSight} based on whether the player's center lies inside this goblin's
     * LOS circle. Operates in world coordinates so the camera transform is irrelevant.
     */
    public void checkLOS() {
        int playerCenterX = gp.Player.WorldX + gp.Player.hitboxDefaultX + (gp.Player.collisionArea.width / 2);
        int playerCenterY = gp.Player.WorldY + gp.Player.hitboxDefaultY + (gp.Player.collisionArea.height / 2);
        int goblinCenterX = WorldX + hitboxDefaultX + (collisionArea.width / 2);
        int goblinCenterY = WorldY + hitboxDefaultY + (collisionArea.height / 2);
        Circle losRange = new Circle(LOSradius, goblinCenterX, goblinCenterY);
        inSight = losRange.intersects(playerCenterX, playerCenterY);
    }

    /**
     * Retrieves the effective direction for rendering the goblin.
     *
     * @return The direction in which the goblin is currently drawn.
     */
    @Override
    protected Direction getEffectiveDirection() {
        return drawDirection; // Default behavior for Player
    }
}

File: web/src/entity/Goblin.ts
import { Direction } from "../app/Direction.ts";
import { Circle } from "../pathfinder/Circle.ts";
import { Camera } from "../render/Camera.ts";
import { CollisionArea } from "../render/CollisionArea.ts";
import { loadImage } from "../render/loadImage.ts";
import { Entity } from "./Entity.ts";
import { Player } from "./Player.ts";

/**
 * Abstract base for hostile entities. Mirrors Goblin.java — same Pascal-cased
 * field naming, same LOS-circle + tile-distance latch for the chase state.
 *
 * Subclasses provide getAction(), which decides per-tick what to do given
 * onPath/inSight + the goblin's current direction. RegularGoblin chases via
 * A* when onPath, otherwise wanders.
 */
export abstract class Goblin extends Entity {
  /** Direction the sprite faces (separate from movement direction so the AI can re-point movement without flipping the sprite). */
  drawDirection: Direction = Direction.DOWN;
  /** Latched chase state — true while pursuing the player. */
  onPath = false;
  /** True iff the player's center sits inside this goblin's LOS circle. */
  inSight = false;
  /** Line-of-sight radius in world pixels. */
  LOSradius = 130;

  protected readonly player: Player;
  protected readonly tileSize: number;

  /** Manhattan tile distance beyond which a chasing goblin gives up if it has lost line of sight. */
  private static readonly DISENGAGE_TILE_DISTANCE = 8;

  constructor(player: Player, worldX: number, worldY: number, tileSize: number) {
    super(worldX, worldY);
    this.player = player;
    this.tileSize = tileSize;
    this.speed = 2;
    this.collisionArea = new CollisionArea(11, 17, 23, 23);
    this.hitboxDefaultX = 11;
    this.hitboxDefaultY = 17;
    this.direction = Direction.DOWN;
  }

  /** Per-tick AI hook. Subclass-specific. */
  abstract getAction(): void;

  /** Loads all eight goblin sprites in parallel. */
  async loadSprites(): Promise<void> {
    const [up1, up2, down1, down2, left1, left2, right1, right2] = await Promise.all([
      loadImage("/goblin/orc_up_1.png"),
      loadImage("/goblin/orc_up_2.png"),
      loadImage("/goblin/orc_down_1.png"),
      loadImage("/goblin/orc_down_2.png"),
      loadImage("/goblin/orc_left_1.png"),
      loadImage("/goblin/orc_left_2.png"),
      loadImage("/goblin/orc_right_1.png"),
      loadImage("/goblin/orc_right_2.png"),
    ]);
    this.up1 = up1;
    this.up2 = up2;
    this.down1 = down1;
    this.down2 = down2;
    this.left1 = left1;
    this.left2 = left2;
    this.right1 = right1;
    this.right2 = right2;
  }

  /**
   * Recomputes LOS and updates the {@code onPath} latch. Engages chase when
   * the player enters the LOS circle; disengages only after losing sight AND
   * drifting more than {@link DISENGAGE_TILE_DISTANCE} tiles away — keeps
   * the goblin pursuing briefly around corners.
   */
  update(): void {
    const xDistance = Math.abs(this.WorldX - this.player.WorldX);
    const yDistance = Math.abs(this.WorldY - this.player.WorldY);
    const tileDistance = Math.floor((xDistance + yDistance) / this.tileSize);

    this.checkLOS();

    if (!this.onPath && this.inSight) {
      this.onPath = true;
    }
    if (this.onPath && !this.inSight && tileDistance > Goblin.DISENGAGE_TILE_DISTANCE) {
      this.onPath = false;
    }

    this.getAction();
  }

  /**
   * Sets {@link inSight} based on whether the player's center lies inside the
   * goblin's LOS circle. Operates in world coordinates so the camera transform
   * is irrelevant.
   */
  checkLOS(): void {
    const playerCenterX = this.player.WorldX + this.player.hitboxDefaultX + this.player.collisionArea.width / 2;
    const playerCenterY = this.player.WorldY + this.player.hitboxDefaultY + this.player.collisionArea.height / 2;
    const goblinCenterX = this.WorldX + this.hitboxDefaultX + this.collisionArea.width / 2;
    const goblinCenterY = this.WorldY + this.hitboxDefaultY + this.collisionArea.height / 2;
    const losRange = new Circle(this.LOSradius, goblinCenterX, goblinCenterY);
    this.inSight = losRange.intersects(playerCenterX, playerCenterY);
  }

  protected override getEffectiveDirection(): Direction {
    return this.drawDirection;
  }

  /** Renders the goblin at its world position through the camera. Skips when fully offscreen. */
  draw(ctx: CanvasRenderingContext2D, camera: Camera): void {
    const sprite = this.getSpriteForDirection();
    if (!sprite) return;

    const screenX = camera.toScreenX(this.WorldX);
    const screenY = camera.toScreenY(this.WorldY);
    const tile = this.tileSize;
    if (screenX + tile <= 0 || screenX >= camera.screenWidth) return;
    if (screenY + tile <= 0 || screenY >= camera.screenHeight) return;

    ctx.drawImage(sprite, screenX, screenY, tile, tile);
  }
}
